File: recipes/claude/04-structure-messages-api-requests/python/messages_requests.py
"""Anthropic Messages API Request Structure & Role Sequencing Harness.

Demonstrates root-level system parameter configuration, multi-turn conversation
history validation, multi-modal content blocks (text, image, PDF document),
tool use blocks, and error handling for malformed request structures.
"""

from __future__ import annotations

import base64
import os
import sys
import time
from typing import Any

import anthropic
from dotenv import load_dotenv

load_dotenv()

API_KEY = os.getenv("ANTHROPIC_API_KEY")
if not API_KEY:
    print("Error: ANTHROPIC_API_KEY environment variable is missing.", file=sys.stderr)
    sys.exit(1)

client = anthropic.Anthropic(api_key=API_KEY)
PRIMARY_MODEL = "claude-3-7-sonnet-20250219"


def run_basic_structured_request() -> anthropic.types.Message:
    """Demonstrate root-level system prompt and alternating message turns."""
    print("\n[1/4] Testing basic request with root system prompt & role alternation...")
    start_time = time.perf_counter()

    # The system prompt is passed at the root of the client call, NEVER inside messages list
    response = client.messages.create(
        model=PRIMARY_MODEL,
        max_tokens=256,
        temperature=0.1,
        system="You are an enterprise system architect. Answer concisely with bulleted points.",
        messages=[
            {
                "role": "user",
                "content": "Why does Anthropic require system instructions outside the messages array?",
            },
            {
                "role": "assistant",
                "content": (
                    "Anthropic isolates system prompts at the root parameter to enforce "
                    "strict conversational boundaries, cache system prefixes efficiently, and "
                    "prevent role-inversion prompt injection vulnerabilities."
                ),
            },
            {
                "role": "user",
                "content": "What happens if a client application attempts to append a role='system' in messages?",
            },
        ],
    )

    elapsed = time.perf_counter() - start_time
    print(f"Status: Success ({elapsed:.2f}s)")
    print(f"Stop Reason: {response.stop_reason}")
    print(f"Input Tokens: {response.usage.input_tokens}, Output Tokens: {response.usage.output_tokens}")
    for block in response.content:
        if block.type == "text":
            print(f"Response: {block.text.strip()}")
    return response


def run_multimodal_content_blocks() -> anthropic.types.Message:
    """Demonstrate multi-modal content blocks (text + 1x1 synthetic PNG image)."""
    print("\n[2/4] Testing multi-modal content blocks (text + image block)...")
    start_time = time.perf_counter()

    # 1x1 transparent PNG payload in base64
    tiny_png_b64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII="

    response = client.messages.create(
        model=PRIMARY_MODEL,
        max_tokens=200,
        temperature=0.0,
        system="Analyze incoming image assets and report technical specifications.",
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "text",
                        "text": "Inspect this image block and confirm receipt.",
                    },
                    {
                        "type": "image",
                        "source": {
                            "type": "base64",
                            "media_type": "image/png",
                            "data": tiny_png_b64,
                        },
                    },
                ],
            }
        ],
    )

    elapsed = time.perf_counter() - start_time
    print(f"Status: Success ({elapsed:.2f}s)")
    for block in response.content:
        if block.type == "text":
            print(f"Response: {block.text.strip()}")
    return response


def run_tool_use_roundtrip() -> anthropic.types.Message:
    """Demonstrate assistant tool_use blocks and subsequent user tool_result response."""
    print("\n[3/4] Testing tool_use and tool_result block sequencing...")
    start_time = time.perf_counter()

    tools: list[dict[str, Any]] = [
        {
            "name": "get_database_schema_version",
            "description": "Returns current production database schema revision.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "cluster_id": {
                        "type": "string",
                        "description": "Identifier for the target database cluster",
                    }
                },
                "required": ["cluster_id"],
            },
        }
    ]

    # Initial query prompting a tool call
    first_turn = client.messages.create(
        model=PRIMARY_MODEL,
        max_tokens=512,
        temperature=0.0,
        system="You are an automated operations agent. Use available tools when requested.",
        tools=tools,
        messages=[
            {
                "role": "user",
                "content": "Check the schema version for database cluster 'pg-primary-us-east-1'.",
            }
        ],
    )

    print(f"Turn 1 stop_reason: {first_turn.stop_reason}")

    tool_call_id = ""
    for block in first_turn.content:
        if block.type == "tool_use":
            tool_call_id = block.id
            print(f"Assistant initiated tool_use: id={block.id}, name={block.name}, input={block.input}")

    if not tool_call_id:
        print("Error: Expected tool_use block from model.")
        return first_turn

    # Execute simulated tool and return result block
    second_turn = client.messages.create(
        model=PRIMARY_MODEL,
        max_tokens=512,
        temperature=0.0,
        system="You are an automated operations agent. Use available tools when requested.",
        tools=tools,
        messages=[
            {
                "role": "user",
                "content": "Check the schema version for database cluster 'pg-primary-us-east-1'.",
            },
            {
                "role": "assistant",
                "content": first_turn.content,
            },
            {
                "role": "user",
                "content": [
                    {
                        "type": "tool_result",
                        "tool_use_id": tool_call_id,
                        "content": '{"cluster_id": "pg-primary-us-east-1", "schema_version": "2026.09.1", "status": "healthy"}',
                    }
                ],
            },
        ],
    )

    elapsed = time.perf_counter() - start_time
    print(f"Turn 2 complete ({elapsed:.2f}s). Final assistant response:")
    for block in second_turn.content:
        if block.type == "text":
            print(f"Response: {block.text.strip()}")
    return second_turn


def run_malformed_structure_probe() -> None:
    """Verify that malformed requests (consecutive user roles or system in messages) are caught."""
    print("\n[4/4] Validating error boundaries on invalid structure...")
    try:
        # Invalid: consecutive user messages without assistant turn
        client.messages.create(
            model=PRIMARY_MODEL,
            max_tokens=100,
            messages=[
                {"role": "user", "content": "Message 1"},
                {"role": "user", "content": "Message 2 without assistant turn"},
            ],
        )
        print("Notice: Consecutive user messages merged or accepted by endpoint.")
    except anthropic.BadRequestError as e:
        print(f"Correctly caught BadRequestError on invalid structure: {e.message}")
    except Exception as e:
        print(f"Caught expected validation error: {type(e).__name__}: {e}")


if __name__ == "__main__":
    print("=================================================================")
    print("Anthropic Messages API Request Structure & Role Suite")
    print("=================================================================")
    try:
        run_basic_structured_request()
        run_multimodal_content_blocks()
        run_tool_use_roundtrip()
        run_malformed_structure_probe()
        print("\nAll request structure tests passed.")
    except Exception as err:
        print(f"\nExecution failed: {err}", file=sys.stderr)
        sys.exit(1)

File: recipes/claude/04-structure-messages-api-requests/typescript/messages_requests.ts
/**
 * Anthropic Messages API Request Structure & Role Sequencing Harness (TypeScript).
 *
 * Demonstrates root-level system parameter configuration, multi-turn role alternation,
 * multi-modal content blocks, tool use handling, and request configuration parameters.
 */

import Anthropic from "@anthropic-ai/sdk";
import * as dotenv from "dotenv";

dotenv.config();

const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) {
  console.error("Error: ANTHROPIC_API_KEY environment variable is not set.");
  process.exit(1);
}

const client = new Anthropic({ apiKey });
const PRIMARY_MODEL = "claude-3-7-sonnet-20250219";

async function runBasicStructuredRequest(): Promise<void> {
  console.log("\n[1/4] Running basic request with root-level system prompt & role alternation...");
  const startTime = Date.now();

  const response = await client.messages.create({
    model: PRIMARY_MODEL,
    max_tokens: 256,
    temperature: 0.1,
    system: "You are an expert distributed systems engineer. Answer concisely.",
    messages: [
      {
        role: "user",
        content: "What is the primary operational distinction between Anthropic's system parameter and the messages array?",
      },
      {
        role: "assistant",
        content:
          "The system parameter is supplied at the root level of the request rather than as a role inside the messages array. This provides dedicated prompt isolation, supports prompt caching, and strictly enforces user and assistant turn alternation.",
      },
      {
        role: "user",
        content: "Summarize the required role alternation pattern in multi-turn dialogues.",
      },
    ],
  });

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`Status: Success (${elapsed}s)`);
  console.log(`Stop Reason: ${response.stop_reason}`);
  console.log(`Tokens - Input: ${response.usage.input_tokens}, Output: ${response.usage.output_tokens}`);

  for (const block of response.content) {
    if (block.type === "text") {
      console.log(`Response: ${block.text.trim()}`);
    }
  }
}

async function runMultimodalContentBlocks(): Promise<void> {
  console.log("\n[2/4] Testing multi-modal content blocks (text + image block)...");
  const startTime = Date.now();

  // 1x1 transparent PNG payload in base64
  const tinyPngBase64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=";

  const response = await client.messages.create({
    model: PRIMARY_MODEL,
    max_tokens: 200,
    temperature: 0.0,
    system: "You are an automated vision asset processor.",
    messages: [
      {
        role: "user",
        content: [
          {
            type: "text",
            text: "Inspect this 1x1 pixel image block and verify schema compliance.",
          },
          {
            type: "image",
            source: {
              type: "base64",
              media_type: "image/png",
              data: tinyPngBase64,
            },
          },
        ],
      },
    ],
  });

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`Status: Success (${elapsed}s)`);
  for (const block of response.content) {
    if (block.type === "text") {
      console.log(`Response: ${block.text.trim()}`);
    }
  }
}

async function runToolUseRoundtrip(): Promise<void> {
  console.log("\n[3/4] Testing tool_use and tool_result block sequencing...");
  const startTime = Date.now();

  const tools: Anthropic.Tool[] = [
    {
      name: "query_service_health",
      description: "Returns health status and uptime metrics for a microservice.",
      input_schema: {
        type: "object",
        properties: {
          service_name: {
            type: "string",
            description: "Target service name to probe",
          },
        },
        required: ["service_name"],
      },
    },
  ];

  // Turn 1: Ask question requiring tool call
  const firstTurn = await client.messages.create({
    model: PRIMARY_MODEL,
    max_tokens: 512,
    system: "You are an automated infrastructure monitoring bot. Call available tools to check status.",
    tools,
    messages: [
      {
        role: "user",
        content: "Check status for service 'auth-gateway-v2'.",
      },
    ],
  });

  let toolCallId = "";
  let toolName = "";
  for (const block of firstTurn.content) {
    if (block.type === "tool_use") {
      toolCallId = block.id;
      toolName = block.name;
      console.log(`Assistant initiated tool call: [${toolName}] id=${toolCallId}`);
    }
  }

  if (!toolCallId) {
    console.error("Expected tool_use block from model.");
    return;
  }

  // Turn 2: Feed back tool_result block
  const secondTurn = await client.messages.create({
    model: PRIMARY_MODEL,
    max_tokens: 512,
    system: "You are an automated infrastructure monitoring bot. Call available tools to check status.",
    tools,
    messages: [
      {
        role: "user",
        content: "Check status for service 'auth-gateway-v2'.",
      },
      {
        role: "assistant",
        content: firstTurn.content,
      },
      {
        role: "user",
        content: [
          {
            type: "tool_result",
            tool_use_id: toolCallId,
            content: JSON.stringify({
              service: "auth-gateway-v2",
              status: "nominal",
              latency_ms: 18,
              uptime_pct: 99.98,
            }),
          },
        ],
      },
    ],
  });

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(2);
  console.log(`Turn 2 complete (${elapsed}s). Final response:`);
  for (const block of secondTurn.content) {
    if (block.type === "text") {
      console.log(`Response: ${block.text.trim()}`);
    }
  }
}

async function runMalformedStructureValidation(): Promise<void> {
  console.log("\n[4/4] Validating error boundaries on malformed request payloads...");
  try {
    // Attempting to pass role='system' directly inside messages array (which is invalid in Anthropic API)
    // @ts-expect-error Intentionally invalid payload to verify API error boundary
    await client.messages.create({
      model: PRIMARY_MODEL,
      max_tokens: 50,
      messages: [
        {
          role: "system",
          content: "This should fail schema validation.",
        },
      ],
    });
    console.log("Notice: Call unexpectedly completed.");
  } catch (err: unknown) {
    if (err instanceof Anthropic.BadRequestError) {
      console.log(`Correctly caught Anthropic.BadRequestError: ${err.message}`);
    } else if (err instanceof Error) {
      console.log(`Caught error boundary check: ${err.name} - ${err.message}`);
    }
  }
}

async function main(): Promise<void> {
  console.log("=================================================================");
  console.log("Anthropic Messages API Request Structure & Role Suite (TS)");
  console.log("=================================================================");
  try {
    await runBasicStructuredRequest();
    await runMultimodalContentBlocks();
    await runToolUseRoundtrip();
    await runMalformedStructureValidation();
    console.log("\nAll TypeScript request structure tests passed.");
  } catch (error) {
    console.error("Test execution failed:", error);
    process.exit(1);
  }
}

main();
